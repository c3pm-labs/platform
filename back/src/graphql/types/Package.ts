import { objectType } from '@nexus/schema';

import { PackageService } from '../../services/PackageService';

export const Versions = objectType({
  name: 'Version',
  definition(t) {
    t.model.description();
    t.model.license();
    t.model.package();
    t.model.readme();
    t.model.publishedAt();
    t.model.version();
  },
});

export const Package = objectType({
  name: 'Package',
  definition(t) {
    t.model.author();
    t.model.name();
    t.model.versions();
    t.field('latest', {
      type: 'Version',
      resolve: async (pkg, _args, ctx) => {
        const packageService = new PackageService(ctx);
        return packageService.getLatestVersion(pkg.name);
      },
    });
  },
});
