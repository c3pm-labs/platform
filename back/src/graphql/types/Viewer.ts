import { objectType } from '@nexus/schema';

export const Viewer = objectType({
  name: 'Viewer',
  definition(t) {
    t.model('User').id();
    t.model('User').email();
    t.model('User').username();
  },
});
