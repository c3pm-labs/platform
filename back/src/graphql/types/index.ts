export * from './User';
export * from './Mutation';
export * from './Query';
export * from './Package';
