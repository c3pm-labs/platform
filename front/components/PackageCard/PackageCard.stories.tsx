import { Meta } from '@storybook/react';

import PackageCard from '.';

const versions = [
  {
    version: '1.2.3', publishedAt: '10/12/20', description: '', readme: '',
  },
  {
    version: '1.1.4', publishedAt: '10/13/20', description: '', readme: '',
  },
  {
    version: '1.0.0', publishedAt: '10/14/20', description: '', readme: '',
  },
  {
    version: '0.9.8', publishedAt: '10/28/20', description: '', readme: '',
  },
  {
    version: '2.0.6', publishedAt: '10/02/20', description: '', readme: '',
  },
  {
    version: '0.3.0', publishedAt: '10/22/20', description: '', readme: '',
  },
];

export default {
  title: 'PackageCard',
  argTypes: {
    name: {
      control: { type: 'text' },
      defaultValue: 'boost',
    },
    author: {
      name: 'author',
      control: { type: 'text' },
      defaultValue: 'Itagiba A.',
    },
    email: {
      name: 'author',
      control: { type: 'text' },
      defaultValue: 'chloe.bourbion@gmail.com',
    },
  },
} as Meta;

interface ArgType {
  name: string;
  author: string;
  email: string;
}

export const Basic = (args: ArgType): JSX.Element => {
  const { name, author, email } = args;
  return (
    <PackageCard packageData={{
      name,
      author: { name: author, id: '', email },
      versions,
      latest: versions[0],
    }}
    />
  );
};
