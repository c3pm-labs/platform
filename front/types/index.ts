export interface Package {
  name: string;
  latest: Version;
  versions: Version[];
  author: User;
  downloads: number;
  tags: string[];
}

export interface Version {
  version: string;
  publishedAt: string;
  description: string;
  readme: string;
  package?: Package;
}

export interface User {
  id: string;
  username: string;
  email: string;
  description: string;
  packages: Package[];
}

export interface Viewer extends User {
  username: string;
  email: string;
}
