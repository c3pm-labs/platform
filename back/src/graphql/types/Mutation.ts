import { mutationType, stringArg } from '@nexus/schema';

import { AuthService } from '../../services/AuthService';
import { UpdateService } from '../../services/UpdateService';

import { Viewer } from './Viewer';
import { User } from './User';

export const Mutation = mutationType({
  definition(t) {
    t.field('register', {
      type: Viewer,
      args: {
        username: stringArg({ required: true }),
        email: stringArg({ required: true }),
        password: stringArg({ required: true }),
      },
      resolve: async (parent, args, ctx) => {
        const authService = new AuthService(ctx);
        return authService.register(args);
      },
    });
    t.field('login', {
      type: Viewer,
      args: {
        login: stringArg({ required: true }),
        password: stringArg({ required: true }),
      },
      resolve: async (parent, args, ctx) => {
        const authService = new AuthService(ctx);
        return authService.login(args);
      },
    });
    t.field('logout', {
      type: Viewer,
      resolve: (parent, args, ctx) => {
        const authService = new AuthService(ctx);
        return authService.logout();
      },
    });
    t.field('update', {
      type: User,
      nullable: true,
      args: {
        id: stringArg(),
        username: stringArg(),
        email: stringArg(),
        description: stringArg(),
      },
      resolve: async (parent, args, ctx) => {
        const updateService = new UpdateService(ctx);
        return updateService.updateUser(args);
      },
    });
  },
});
