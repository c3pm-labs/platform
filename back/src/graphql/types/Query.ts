import { queryType, stringArg } from '@nexus/schema';

import { AuthService } from '../../services/AuthService';
import { UserService } from '../../services/UserService';
import { PackageService } from '../../services/PackageService';

import { Viewer } from './Viewer';
import { User } from './User';
import { Package, Versions } from './Package';

export const Query = queryType({
  definition(t) {
    t.field('viewer', {
      type: Viewer,
      resolve: async (parent, args, ctx) => {
        const authService = new AuthService(ctx);
        return authService.viewer();
      },
    });
    t.field('user', {
      type: User,
      nullable: true,
      args: {
        username: stringArg(),
        id: stringArg(),
      },
      resolve: async (parent, args, ctx) => {
        const userService = new UserService(ctx);
        return userService.findOne(args);
      },
    });
    t.list.field('search', {
      type: Package,
      args: {
        keyword: stringArg({ required: true }),
      },
      resolve: async (parent, args, ctx) => {
        const packageService = new PackageService(ctx);
        return packageService.search(args.keyword);
      },
    });
    t.field('package', {
      type: Package,
      args: {
        name: stringArg({ required: true }),
      },
      resolve: async (parent, args, ctx) => {
        const packageService = new PackageService(ctx);
        return packageService.getPackage(args.name);
      },
    });
    t.field('version', {
      type: Versions,
      args: {
        packageName: stringArg({ required: true }),
        version: stringArg({ required: false }),
      },
      resolve: async (parent, args, ctx) => {
        const packageService = new PackageService(ctx);
        return packageService.getVersionOrLatest(args.packageName, args.version);
      },
    });
  },
});
