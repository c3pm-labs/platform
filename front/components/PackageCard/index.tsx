import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import { Typography } from '@material-ui/core';
import { useTranslation } from 'next-i18next';
import { useRouter } from 'next/router';
import clsx from 'clsx';

import Avatar from '../Avatar';
import { Package } from '../../types';
import TextLink from '../TextLink';

export interface PackageCardProps {
  packageData: Package;
  discover?: string;
}

const useStyles = makeStyles((theme: Theme) => createStyles({
  container: {
    minHeight: '8rem',
    padding: theme.spacing(3),
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-between',
    background: '#FFFFFF',
    boxShadow: '0px 4px 15px rgba(0, 0, 0, 0.12)',
    borderRadius: theme.shape.borderRadius,
    [theme.breakpoints.up('sm')]: {
      minWidth: '600px',
    },
    [theme.breakpoints.down('xs')]: {
      maxWidth: '100%',
    },
  },
  top: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'flex-start',
    '& h5': {
      fontSize: 18,
      fontWeight: 'bold',
    },
    marginBottom: theme.spacing(1),
  },
  bottom: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-between',
    alignItems: 'flex-start',
    marginTop: theme.spacing(1),
  },
  update: {
    width: '100%',
    textAlign: 'right',
    fontSize: 10,
  },
  description: {
    lineHeight: 1.2,
    fontSize: 14,
    fontWeight: 400,
  },
  avatar: {
    width: 30,
    height: 30,
    marginRight: theme.spacing(1),
  },
  name: {
    fontSize: 14,
  },
  tagsContainer: {
    display: 'flex',
    flexWrap: 'wrap',
    width: '100%',
    justifyContent: 'flex-end',
  },
  tag: {
    fontSize: 12,
    fontWeight: 'bold',
    marginRight: 5,
    color: theme.palette.text.primary,
    background: 'rgba(0,184,230, 0.3)', // main with opacity
    borderRadius: theme.shape.borderRadius,
    padding: '3px 6px',
  },
  downloads: {
    borderRadius: 7,
    padding: '2px 8px',
    color: 'white',
  },
}));

function PackageCard({ packageData, discover = undefined }: PackageCardProps): JSX.Element {
  const classes = useStyles({ packageData });
  const router = useRouter();
  const { t } = useTranslation('common');
  const {
    name, author, tags, latest, downloads,
  } = packageData;
  const date = new Date(latest.publishedAt);
  const url = `/package/${name}/${latest.version}`;

  return (
    <div className={classes.container} data-testid={`packageCard-${name}`}>
      <div className={classes.top}>
        <Typography variant="h5" data-testid={`packageCard-${name}-name`}>
          <TextLink href={url} as={url}>
            {name}
            <span style={{ fontSize: 14 }}>
              {' v'}
              {latest.version}
            </span>
          </TextLink>
        </Typography>
        <div className={discover ? clsx(classes.downloads, discover) : classes.description}>
          {!discover ? 'ðŸ”¥' : null}
          {downloads}
          {discover
            ? (
              <span style={{ fontSize: 10 }}>
                {t('downloads')}
              </span>
            ) : null}
        </div>
      </div>
      <div style={{ display: 'flex', flexWrap: 'wrap', overflowWrap: 'break-word' }}>
        <Typography color="textPrimary" variant="body1" style={{ width: '100%' }}>{latest.description}</Typography>
      </div>
      <div className={classes.tagsContainer}>
        {tags.slice(0, 3)?.map((tag) => <span key={`tag-${tag}-${name}`} className={classes.tag}>{tag}</span>)}
      </div>
      <div className={classes.bottom}>
        <Avatar
          user={author}
          classes={{ picture: classes.avatar, name: classes.name }}
        />
        <Typography variant="body2" className={classes.update}>
          {t('lastUpdated')}
          &nbsp;
          {router.locale === 'en' ? date.toDateString() : date.toLocaleDateString()}
        </Typography>
      </div>
    </div>
  );
}

export default PackageCard;
