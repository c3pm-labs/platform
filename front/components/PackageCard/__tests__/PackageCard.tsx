import { render } from '@testing-library/react';
import { Package } from 'types';

import PackageCard from '../index';

test('PackageCard', () => {
  const mockPackage = {
    name: '',
    latest: {
      version: '0.2',
      publishedAt: '10:04:1999',
      description: 'hot fix',
      readme: 'blablabla',
    },
    versions: [
      {
        version: '0.2',
        publishedAt: '10:04:1999',
        description: 'hot fix',
        readme: 'blablabla',
      },
      {
        version: '0.2',
        publishedAt: '10:04:1999',
        description: 'hot fix',
        readme: 'blablabla',
      },
    ],
    author: {
      id: '12345',
      username: 'Bob',
      email: 'bob.dylan@gmail.com',
      description: 'singer',
      packages: [],
    },
  };

  const { container } = render(
    <PackageCard packageData={mockPackage as Package} />,
  );

  expect(container).toMatchSnapshot();
});
