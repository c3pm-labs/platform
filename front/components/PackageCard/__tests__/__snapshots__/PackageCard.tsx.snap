// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackageCard 1`] = `
<div>
  <div
    class="makeStyles-container-1"
    data-testid="packageCard-"
  >
    <div
      class="makeStyles-top-2"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5"
        data-testid="packageCard--name"
      >
        <a
          class="MuiTypography-root MuiLink-root MuiLink-underlineAlways MuiTypography-colorPrimary"
          href="/package/"
        />
      </h5>
      <h5
        class="MuiTypography-root MuiTypography-h5 MuiTypography-colorPrimary"
        data-testid="packageCard--version"
      >
        0.2
      </h5>
    </div>
    <div
      style="display: flex; flex-wrap: wrap;"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextPrimary"
      >
        hot fix
      </p>
    </div>
    <p
      class="MuiTypography-root makeStyles-description-5 MuiTypography-body1 MuiTypography-colorTextPrimary"
    >
      hot fix
    </p>
    <div
      class="makeStyles-bottom-3"
    >
      <div
        class="makeStyles-container-8"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle makeStyles-picture-9 makeStyles-picture-11 makeStyles-avatar-6 MuiAvatar-colorDefault"
          data-testid="user-avatar"
        >
          B
        </div>
        <p
          class="MuiTypography-root makeStyles-name-10 makeStyles-name-7 MuiTypography-body1"
        >
          Bob
        </p>
      </div>
      <p
        class="MuiTypography-root makeStyles-update-4 MuiTypography-body2"
      >
        Last updated on theÂ 
        Fri Jan 01 1999
      </p>
    </div>
  </div>
</div>
`;
